golang_src/cmd/worker.go

```diff
@@ -0,0 +1,137 @@
+package main
+
+import (
+	"crypto/rand"
+	"fmt"
+	"github.com/cerberus98/babymailgun"
+	"github.com/spf13/viper"
+	"log"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+)
+
+type WorkerConfig struct {
+	Host         string
+	Port         string
+	DatabaseName string
+	WorkerSleep  int
+	WorkerPool   int
+	MailHost     string
+	MailPort     string
+	AdminEmail   string
+}
+
+// Create a UUID4. Source implementation here: https://groups.google.com/forum/#!msg/golang-nuts/d0nF_k4dSx4/rPGgfXv6QCoJ
+func uuid() string {
+	b := make([]byte, 16)
+	rand.Read(b)
+	b[6] = (b[6] & 0x0f) | 0x40
+	b[8] = (b[8] & 0x3f) | 0x80
+	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:])
+}
+
+func processingLoop(cfg *WorkerConfig, wg *sync.WaitGroup, quit <-chan bool) {
+	// Fetch emails ready to be sent
+	// Try to send them
+	// Update the datastore
+	// Go back to sleep
+	log.Println("Starting processing loop...")
+	workerId := uuid()
+	defer wg.Done()
+loop:
+	for {
+		mongoClient := babymailgun.MongoClient{Host: cfg.Host, Port: cfg.Port, DatabaseName: cfg.DatabaseName}
+		select {
+		case <-quit:
+			log.Printf("Worker goroutine received quit")
+			break loop
+		default:
+			log.Println("Waking up and looking for emails to send")
+		}
+		email, err := mongoClient.FetchReadyEmail(workerId)
+		if err == nil {
+			log.Printf("Got email %s Worker ID: %s\n", email.ID, workerId)
+			// Try to send the email
+			mc := babymailgun.MailConfig{MailHost: cfg.MailHost, MailPort: cfg.MailPort, AdminEmail: cfg.AdminEmail}
+			if err = babymailgun.SendMail(&mc, email); err != nil {
+				fmt.Println("Email sending failed ", err)
+				// TODO Set the reason on the update document
+				// TODO Some errors are definitely catastrophic
+			}
+			// Process the email response to see if any recipients failed and create
+			// and update document. Additionally, "release" the email by setting the
+			// worker_id to nil
+
+			log.Printf("Updating and releasing email %s\n", email.ID)
+			if err := mongoClient.UpdateEmail(email); err != nil {
```

Comment:
`UpdateEmail` only resets the `worker_id` to `nil` for the email in the database. That makes it look as if it has never been delivered. So it can be picked up for delivery again by `FetchReadyEmail`, is not it?

Reply:
Also WIP. There's a comment in there to generate a complete update document, and also parse the server response to record addresses that failed/bounced/whatever. For now, I was just testing the concurrency semantics to make sure findAndModify()/Change works the way I expected. In other words, the code is fetching an email, ensuring no others are also fetching that email, and then "releasing" it by resetting the worker_id field